### PR TITLE
refactor(@desktop/chat-messages): add/remove reactions

### PR DESCRIPTION
**Pull request by saledjenic**
Add/remove reactions added for messages from a message list and for messages displayed in the pinned messages list.

### CODE REVIEW

**Review by borismelnik — APPROVED** [review]

File: src/app/modules/main/chat_section/chat_content/messages/module.nim [hunk 1]
```diff
@@ -92,7 +96,14 @@ method newMessagesLoaded*(self: Module, messages: seq[MessageDto], reactions: se
 
     for r in reactions:
       if(r.messageId == m.id):
-        item.addReaction(r.emojiId, m.`from`, r.id)
+        var emojiIdAsEnum: EmojiId
+        if(message_reaction_item.toEmojiIdAsEnum(r.emojiId, emojiIdAsEnum)):
+          let userWhoAddedThisReaction = self.controller.getContactById(r.`from`)
+          let amIReactedWithThisEmoji = userWhoAddedThisReaction.id == singletonInstance.userProfile.getPubKey()
```
Comment (jrainville) [inline src/app/modules/main/chat_section/chat_content/messages/module.nim:102]: Not a big deal, but this name isn't grammatically correct
It should be either `amITheOneWhoReactedWithThisEmoji`
or `didIReactWithThisEmoji`
Worst case, `isCurrentUser` works as well

File: src/app/modules/main/chat_section/chat_content/module.nim [hunk 2]
```diff
@@ -151,8 +152,14 @@ proc buildPinnedMessageItem(self: Module, messageId: string, item: var pinned_ms
 
   for r in reactions:
     if(r.messageId == m.id):
-      # m.`from` should be replaced by appropriate ens/alias when we have that part refactored
-      item.addReaction(r.emojiId, m.`from`, r.id)
+      var emojiIdAsEnum: pinned_msg_reaction_item.EmojiId
+      if(pinned_msg_reaction_item.toEmojiIdAsEnum(r.emojiId, emojiIdAsEnum)):
+        let userWhoAddedThisReaction = self.controller.getContactById(r.`from`)
+        let amIReactedWithThisEmoji = userWhoAddedThisReaction.id == singletonInstance.userProfile.getPubKey()
```
Comment (jrainville) [inline src/app/modules/main/chat_section/chat_content/module.nim:158]: Same here

File: src/app/modules/shared_models/message_reaction_item.nim [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      return false
+  return true
+
+proc getReactionId*(self: MessageReactionItem, userPublicKey: string): string = 
+  for r in self.reactions:
+    if (r.publicKey == userPublicKey):
+      return r.reactionId
+  return ""
+
+proc addReaction*(self: var MessageReactionItem, amIReactedWithThisEmoji: bool, userPublicKey: string,
```
Comment (jrainville) [inline src/app/modules/shared_models/message_reaction_item.nim:73]: Same here, but I'll stop commenting on each one. Just change them using `ctrl+shift+f` if you feel like it

File: src/app/modules/shared_models/message_reaction_model.nim [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    if(self.reactionItemWithEmojiIdExists(emojiId)):
+      let ind = self.getIndexOfTheItemWithEmojiId(emojiId)
+      if(ind == -1):
+        return
+      self.items[ind].addReaction(amIReactedWithThisEmoji, userPublicKey, userDisplayName, reactionId)
+      let index = self.createIndex(ind, 0, nil)
+      self.dataChanged(index, index, @[]) # all roles
+    else:
+      let parentModelIndex = newQModelIndex()
+      defer: parentModelIndex.delete
```
Comment (jrainville) [inline src/app/modules/shared_models/message_reaction_model.nim:120]: What does this do?

**Review by jrainville — APPROVED** [review]
Good job. Feel free to rename the property if you want. I can't promise that I won't change it myself someday if you don't, because it's gonna taunt me anytime I see it hahaha.

File: src/app/modules/shared_models/message_reaction_model.nim (hunk 4 above)
Comment (saledjenic) [reply src/app/modules/shared_models/message_reaction_model.nim:120]: @jrainville if you refer to this line `defer: parentModelIndex.delete`, it actually does deferred deletion of `newQModelIndex` and avoids memory leak caused by creating `new QModelIndex` on the c++ side.

**Comment by saledjenic** [comment]
> Feel free to rename the property if you want. I can't promise that I won't change it myself someday if you don't, because it's gonna taunt me anytime I see it hahaha.

@jrainville yes, sure, I will update it. Will use this variant `didIReactWithThisEmoji`, actually every other option I tried looked too long to me and was shorting it more and more and end up with `userWhoAddedThisReaction`. :D

**Pull request merged**